modules/table/src/test/java/org/apache/ignite/internal/table/distributed/replicator/IndexBuilderTxRwOperationTrackerTest.java

```diff
@@ -0,0 +1,171 @@
+        assertTrue(awaitFuture1.isDone());
+    }
+
+    @Test
+    void testRejectTxRwOperation() {
+        tracker.updateMinAllowedCatalogVersionForStartOperation(1);
+
+        assertThrows(RejectTxRwOperationException.class, () -> tracker.incrementOperationCount(0));
+
+        assertDoesNotThrow(() -> tracker.incrementOperationCount(1));
+        assertDoesNotThrow(() -> tracker.incrementOperationCount(2));
+    }
+
+    @Test
+    void testConcurrentUpdateMinAllowedAndAwaitCompleteOperations0() {
+        runRace(
+                () -> {
+                    for (int i = 0; i < 100; i++) {
+                        tracker.updateMinAllowedCatalogVersionForStartOperation(i);
+                    }
+                },
+                () -> {
+                    for (int i = 0; i < 100; i++) {
+                        assertThat(tracker.awaitCompleteTxRwOperations(i), willSucceedFast());
+                    }
+                },
+                () -> {
+                    for (int i = 0; i < 100; i += 5) {
+                        assertThat(tracker.awaitCompleteTxRwOperations(i), willSucceedFast());
+                    }
+                },
+                () -> {
+                    for (int i = 0; i < 100; i += 10) {
+                        assertThat(tracker.awaitCompleteTxRwOperations(i), willSucceedFast());
+                    }
+                }
+        );
+    }
+
+    @Test
+    void testConcurrentUpdateMinAllowedAndAwaitCompleteOperations1() {
+        for (int i = 0; i < REPEATS; i++) {
+            int catalogVersion = i;
+
+            runRace(
+                    () -> tracker.updateMinAllowedCatalogVersionForStartOperation(catalogVersion),
+                    () -> assertThat(tracker.awaitCompleteTxRwOperations(catalogVersion), willSucceedFast())
+            );
+        }
+    }
+
+    @Test
+    void testConcurrentIncrementAndDecrementOperationCount0() {
+        runRace(
+                () -> {
+                    for (int i = 0; i < 100; i++) {
+                        tracker.incrementOperationCount(0);
+                        tracker.decrementOperationCount(0);
+                    }
+                },
+                () -> {
+                    for (int i = 0; i < 100; i++) {
+                        tracker.incrementOperationCount(0);
+                        tracker.decrementOperationCount(0);
+                    }
+                }
+        );
+    }
+
+    @Test
+    void testConcurrentIncrementAndDecrementOperationCount1() {
+        for (int i = 0; i < REPEATS; i++) {
+            runRace(
+                    () -> {
+                        tracker.incrementOperationCount(0);
+                        tracker.decrementOperationCount(0);
+                    },
+                    () -> {
+                        tracker.incrementOperationCount(0);
+                        tracker.decrementOperationCount(0);
+                    }
+            );
+        }
+    }
+}
```

Comment:
Probably the most interesting test scenarios involve interaction between incrementing/decrementing, awaiting and raising the limit, all in the same scenario. How about adding such tests?

Reply:
I think there are enough parallel tests for now, they helped me correct several errors after implementation.